File: csharp/Pokedex.Core/Models/Pokemon.cs
namespace Pokedex.Core.Models;

public class Pokemon
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string FrenchName { get; set; } = string.Empty;
    public string DisplayName => string.IsNullOrEmpty(FrenchName) ? Name : FrenchName;
    public List<string> Types { get; set; } = new();
    public string Description { get; set; } = string.Empty;
    public List<EvolutionChain> Evolutions { get; set; } = new();
    public EvolutionNode? EvolutionTree { get; set; }
    public int Height { get; set; }
    public int Weight { get; set; }
    public List<Stats> Stats { get; set; } = new();
    public List<Ability> Abilities { get; set; } = new();
    public string Genus { get; set; } = string.Empty;
    public int CaptureRate { get; set; }
    public int BaseExperience { get; set; }
    public bool IsLegendary { get; set; }
    public bool IsMythical { get; set; }
    public List<string> Weaknesses { get; set; } = new();
    public List<string> Resistances { get; set; } = new();
    public string Habitat { get; set; } = string.Empty;
    public string Generation { get; set; } = string.Empty;
    public string CryUrl { get; set; } = string.Empty;
}

public class EvolutionChain
{
    public string Name { get; set; } = string.Empty;
    public string FrenchName { get; set; } = string.Empty;
    public int Level { get; set; }
    public string Trigger { get; set; } = string.Empty;
}

public class EvolutionNode
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string FrenchName { get; set; } = string.Empty;
    public string DisplayName => string.IsNullOrEmpty(FrenchName) ? Name : FrenchName;
    public int MinLevel { get; set; }
    public string Trigger { get; set; } = string.Empty;
    public List<EvolutionNode> Children { get; set; } = new();
}

public class Stats
{
    public string Name { get; set; } = string.Empty;
    public int Value { get; set; }
}

public class Ability
{
    public string Name { get; set; } = string.Empty;
    public string FrenchName { get; set; } = string.Empty;
    public string DisplayName => string.IsNullOrEmpty(FrenchName) ? Name : FrenchName;
    public bool IsHidden { get; set; }
}

public class PokemonFormInfo
{
    public string Name { get; set; } = string.Empty;
    public string ImageUrl { get; set; } = string.Empty;
    public List<string> AlternateImageUrls { get; set; } = new();
    public List<string> Types { get; set; } = new();
    public int FormId { get; set; }
}

File: web/src/types/pokemon.ts
export interface NamedResource {
  name: string
  url: string
}

export interface LocalizedName {
  name: string
  language: NamedResource
}

export interface PokemonListItem {
  name: string
  url: string
}

export interface PokemonListResponse {
  count: number
  next: string | null
  results: PokemonListItem[]
}

export interface PokemonApi {
  id: number
  name: string
  height: number
  weight: number
  base_experience: number
  types: { type: NamedResource }[]
  stats: { base_stat: number; stat: NamedResource }[]
  abilities: {
    ability: NamedResource
    is_hidden: boolean
    slot: number
  }[]
  sprites: {
    front_default: string | null
    front_shiny: string | null
    other: {
      'official-artwork': {
        front_default: string | null
        front_shiny: string | null
      }
    }
  }
  cries?: { latest?: string }
}

export interface PokemonSpecies {
  names: LocalizedName[]
  genera: { genus: string; language: NamedResource }[]
  flavor_text_entries: {
    flavor_text: string
    language: NamedResource
  }[]
  evolution_chain: { url: string }
  capture_rate: number
  is_legendary: boolean
  is_mythical: boolean
  habitat?: { name: string } | null
  generation?: { name: string } | null
}

export interface EvolutionDetail {
  min_level: number | null
  trigger: NamedResource | null
  item: NamedResource | null
}

export interface EvolutionChainLink {
  species: NamedResource
  evolution_details: EvolutionDetail[]
  evolves_to: EvolutionChainLink[]
}

export interface EvolutionChainResponse {
  chain: EvolutionChainLink
}

export interface AbilityApi {
  names: LocalizedName[]
}

export interface PokemonSummary {
  id: number
  name: string
  displayName: string
}

export interface PokemonDetail {
  id: number
  name: string
  displayName: string
  genus: string
  description: string
  types: string[]
  height: number
  weight: number
  captureRate: number
  baseExperience: number
  isLegendary: boolean
  isMythical: boolean
  habitat: string
  generation: string
  cryUrl: string
  weaknesses: string[]
  resistances: string[]
  stats: { name: string; value: number }[]
  abilities: { name: string; isHidden: boolean }[]
  evolutionTree: EvolutionNode | null
  forms: PokemonForm[]
  spriteUrl: string
  shinySpriteUrl: string
}

export interface PokemonForm {
  name: string
  imageUrl: string
}

export interface EvolutionNode {
  name: string
  displayName: string
  id: number
  trigger: string
  children: EvolutionNode[]
}
